In [115]:
import pandas as pd
from datetime import datetime
import warnings
warnings.filterwarnings("ignore")

# Работа с Pandas

Загрузите ваш DataFrame в переменную df и выведете на экран первые пять строк 

In [99]:
df = pd.read_csv('googleplaystore.csv')
df.head(5)

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  Reviews  Size     Installs  Type  Price Content Rating  \
0     159   19M      10,000+  Free    0.0       Everyone   
1     967   14M     500,000+  Free    0.0       Everyone   
2   87510  8.7M   5,000,000+  Free    0.0       Everyone   
3  215644   25M  50,000,000+  Free    0.0           Teen   
4     967  2.8M     100,000+  Free    0.0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

Отобразите количество строк и столбцов в DataFrame

In [100]:
df.shape

(10841, 13)

# PRO часть (Не обязательна к выполнению, но желательна :))


Измените названия всех столбцов приведя их к нижнему регистру

In [101]:
df.columns = [col.lower().replace(' ', '_') for col in df.columns]
df.columns

Index(['app', 'category', 'rating', 'reviews', 'size', 'installs', 'type',
       'price', 'content_rating', 'genres', 'last_updated', 'current_ver',
       'android_ver'],
      dtype='object')

Выведите основные статистики по категориальным столбцам. Напишите выводы о них 

In [102]:
df.describe(include='object')

app category reviews                size    installs   type  \
count    10841    10841   10841               10841       10841  10840   
unique    9660       34    6002                 462          22      3   
top     ROBLOX   FAMILY       0  Varies with device  1,000,000+   Free   
freq         9     1972     596                1695        1579  10039   

       content_rating genres    last_updated         current_ver android_ver  
count           10840  10841           10841               10833       10838  
unique              6    120            1378                2832          33  
top          Everyone  Tools  August 3, 2018  Varies with device  4.1 and up  
freq             8714    842             326                1459        2451

count - число элементов<br>
unique - число уникальных элементов<br>
top - какой элемент встречался чаще всего<br>
freq - сколько раз встречался этот элемент<br>

### Вывод:
- `app` имеет больше всг=его уникальных элементов;
- Самое частовстречающееся приложение - ROBLOX
- 1579 приложений скачали более миллиона раз

Выведите на экран часть датафрейма, который соответствует условию: `приложение из категории искусства и дизайна` И `с рейтингом больше 4.5`

In [103]:
df['category'].unique()

array(['ART_AND_DESIGN', 'AUTO_AND_VEHICLES', 'BEAUTY',
       'BOOKS_AND_REFERENCE', 'BUSINESS', 'COMICS', 'COMMUNICATION',
       'DATING', 'EDUCATION', 'ENTERTAINMENT', 'EVENTS', 'FINANCE',
       'FOOD_AND_DRINK', 'HEALTH_AND_FITNESS', 'HOUSE_AND_HOME',
       'LIBRARIES_AND_DEMO', 'LIFESTYLE', 'GAME', 'FAMILY', 'MEDICAL',
       'SOCIAL', 'SHOPPING', 'PHOTOGRAPHY', 'SPORTS', 'TRAVEL_AND_LOCAL',
       'TOOLS', 'PERSONALIZATION', 'PRODUCTIVITY', 'PARENTING', 'WEATHER',
       'VIDEO_PLAYERS', 'NEWS_AND_MAGAZINES', 'MAPS_AND_NAVIGATION',
       '1.9'], dtype=object)

In [104]:
df.loc[(df['category'] == 'ART_AND_DESIGN') & (df['rating'] > 4.5), ['app']]

app
2     U Launcher Lite – FREE Live Cool Themes, Hide ...
9                         Kids Paint Free - Drawing Fun
13                                Mandala Coloring Book
16         Photo Designer - Write your name with shapes
19                                         ibis Paint X
22              Superheroes Wallpapers | 4K Backgrounds
24                          HD Mickey Minnie Wallpapers
25                           Harley Quinn wallpapers HD
26                        Colorfit - Drawing & Coloring
34                                      I Creative Idea
36       UNICORN - Color By Number & Pixel Art Coloring
38                       PIP Camera - PIP Collage Maker
45    Canva: Poster, banner, card maker & graphic de...
46    Install images with music to make video withou...
3982                                  Cardi B Wallpaper
4749  X Launcher: With OS11 Style Theme & Control Ce...
4755  X Launcher Pro: PhoneX Theme, OS11 Control Center
4759  X Launcher Pro - IOS Style Theme & Control Center
4764     X Launcher Prime: With OS Style Theme & No Ads
5254                            AJ Styles HD Wallpapers
6930               Fantasy theme dark bw black building
8888            Spring flowers theme couleurs d t space

Посчитатайте какое количество игр содержится в рейтинге, используя параметр разбиения (bins) равным 100. Выведите на экран топ-10 игр по количеству в рейтинге. Напишите краткий вывод.
То есть сначала отфильтровываете лишь игры из всего датафрейма, после чего применяете value_counts к столбцу с рейтингом используя параметр, который указан выше. И выводите первые 10 строк отсортированных по убыванию.

In [105]:
df.loc[df['category'] == 'GAME', 'rating'].value_counts(bins=100).head(10)

(4.48, 4.52]    174
(4.36, 4.4]     166
(4.28, 4.32]    151
(4.56, 4.6]     121
(4.16, 4.2]     115
(4.08, 4.12]     72
(4.68, 4.72]     67
(3.96, 4.0]      55
(3.76, 3.8]      41
(3.88, 3.92]     35
Name: count, dtype: int64

<b>Выводы:</b><br>
Наибольшее количество игр расположено в интервале с рейтингом около 4.5.<br>
Игр с низким рейтингом нет.

Создайте функцию , которая будет применена к столбцу рейтинг методом `apply()`.
Функция должна принимать строку с рейтингом и возвращать:
 - High rating, если рейтинг больше или равен 4.5
 - Middle rating, если рейтинг между 3.8 и 4.5
 - Low rating, если рейтинг меньше 3.8 
Результат работы функции запишите в новый столбец с названием `categorical_rating`

In [106]:
def make_rating(rating):
    if rating >= 4.5:
        return 'high rating'
    elif rating >= 3.8 and rating < 4.5:
        return 'middle rating'
    else:
        return 'categorial rating'

In [107]:
df['rating'].apply(make_rating)

0            middle rating
1            middle rating
2              high rating
3              high rating
4            middle rating
               ...        
10836          high rating
10837          high rating
10838    categorial rating
10839          high rating
10840          high rating
Name: rating, Length: 10841, dtype: object

In [108]:
df.loc[df['category'] == '1.9']

app category  rating reviews  \
10472  Life Made WI-Fi Touchscreen Photo Frame      1.9    19.0    3.0M   

         size installs type  price content_rating             genres  \
10472  1,000+     Free    0    0.0            NaN  February 11, 2018   

      last_updated current_ver android_ver  
10472       1.0.19  4.0 and up         NaN

In [109]:
df.drop(index=10472, inplace=True)

In [110]:
df.reset_index(inplace=True)

In [111]:
df.drop(columns='index', inplace=True)

Выведите кол-во уникальных значений в столбцах отсортированных по убыванию на экран:

In [112]:
sorted(df.category.unique(), reverse=True)

['WEATHER',
 'VIDEO_PLAYERS',
 'TRAVEL_AND_LOCAL',
 'TOOLS',
 'SPORTS',
 'SOCIAL',
 'SHOPPING',
 'PRODUCTIVITY',
 'PHOTOGRAPHY',
 'PERSONALIZATION',
 'PARENTING',
 'NEWS_AND_MAGAZINES',
 'MEDICAL',
 'MAPS_AND_NAVIGATION',
 'LIFESTYLE',
 'LIBRARIES_AND_DEMO',
 'HOUSE_AND_HOME',
 'HEALTH_AND_FITNESS',
 'GAME',
 'FOOD_AND_DRINK',
 'FINANCE',
 'FAMILY',
 'EVENTS',
 'ENTERTAINMENT',
 'EDUCATION',
 'DATING',
 'COMMUNICATION',
 'COMICS',
 'BUSINESS',
 'BOOKS_AND_REFERENCE',
 'BEAUTY',
 'AUTO_AND_VEHICLES',
 'ART_AND_DESIGN']

__Super-Hard задания__

Решите задачу, где мы создавали свою собственную функцию, но вместо функции `def` используйте `лямбда-функцию` внутри `apply()`

In [113]:
df['rating'].apply(lambda x: 'high rating' if x >= 4.5 else 'middle rating' if x >= 3.8 and x < 4.5 else 'categorial rating')

0            middle rating
1            middle rating
2              high rating
3              high rating
4            middle rating
               ...        
10835          high rating
10836          high rating
10837    categorial rating
10838          high rating
10839          high rating
Name: rating, Length: 10840, dtype: object

Изучите функцию `transform` в документации Pandas. Примените её к датафрейму, который будет заранее сгруппирован по столбцу `category`. С помощью фукнции `transform` найдите средний `rating` по каждой категории и запишите в столбец с названием `mean_cat_rating`. Кол-во строк в получившемся новом столбец должно быть таким же, как и в исходном датафрейме

In [114]:
df['mean_cat_rating'] = df.groupby('category')['rating'].transform('mean')
df

app             category  \
0         Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                    Coloring book moana       ART_AND_DESIGN   
2      U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                  Sketch - Draw & Paint       ART_AND_DESIGN   
4                  Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                  ...                  ...   
10835                                   Sya9a Maroc - FR               FAMILY   
10836                   Fr. Mike Schmitz Audio Teachings               FAMILY   
10837                             Parkinson Exercices FR              MEDICAL   
10838                      The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10839      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       rating reviews                size     installs  type  price  \
0         4.1     159                 19M      10,000+  Free    0.0   
1         3.9     967                 14M     500,000+  Free    0.0   
2         4.7   87510                8.7M   5,000,000+  Free    0.0   
3         4.5  215644                 25M  50,000,000+  Free    0.0   
4         4.3     967                2.8M     100,000+  Free    0.0   
...       ...     ...                 ...          ...   ...    ...   
10835     4.5      38                 53M       5,000+  Free    0.0   
10836     5.0       4                3.6M         100+  Free    0.0   
10837     NaN       3                9.5M       1,000+  Free    0.0   
10838     4.5     114  Varies with device       1,000+  Free    0.0   
10839     4.5  398307                 19M  10,000,000+  Free    0.0   

      content_rating                     genres      last_updated  \
0           Everyone               Art & Design   January 7, 2018   
1           Everyone  Art & Design;Pretend Play  January 15, 2018   
2           Everyone               Art & Design    August 1, 2018   
3               Teen               Art & Design      June 8, 2018   
4           Everyone    Art & Design;Creativity     June 20, 2018   
...              ...                        ...               ...   
10835       Everyone                  Education     July 25, 2017   
10836       Everyone                  Education      July 6, 2018   
10837       Everyone                    Medical  January 20, 2017   
10838     Mature 17+          Books & Reference  January 19, 2015   
10839       Everyone                  Lifestyle     July 25, 2018   

              current_ver         android_ver  mean_cat_rating  
0                   1.0.0        4.0.3 and up         4.358065  
1                   2.0.0        4.0.3 and up         4.358065  
2                   1.2.4        4.0.3 and up         4.358065  
3      Varies with device          4.2 and up         4.358065  
4                     1.1          4.4 and up         4.358065  
...                   ...                 ...              ...  
10835                1.48          4.1 and up         4.192272  
10836                 1.0          4.1 and up         4.192272  
10837                 1.0          2.2 and up         4.189143  
10838  Varies with device  Varies with device         4.346067  
10839  Varies with device  Varies with device         4.094904  

[10840 rows x 14 columns]

Создайте функцию , которая будет применена к столбцу `last updated` и возвращать из него число месяца, если число месяца вернуть из строки невозможно, то верните вместо этого строку с значением `miss_date`.\
Запишите полученный столбец в датафрейм , название столбцу дайте `day_of_update`.

Если вы столкнётесь с некоторой ошибкой , которую вам выдаст Pandas, то изучите конструкцию Python `try` \ `except`.\
Вот псевдокод , который поможет лучше разобраться с этой конструкцией:

In [116]:
def extract_month(date_string):
    try:
        # Пробуем преобразовать строку в дату
        date_obj = datetime.strptime(date_string, '%Y-%m-%d')
        # Возвращаем номер месяца
        return date_obj.month
    except ValueError:
        # Если возникла ошибка, возвращаем 'miss_date'
        return 'miss_date'

# Применяем функцию к столбцу 'last updated'
df['day_of_update'] = df['last_updated'].apply(extract_month)

In [117]:
df

app             category  \
0         Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                    Coloring book moana       ART_AND_DESIGN   
2      U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                  Sketch - Draw & Paint       ART_AND_DESIGN   
4                  Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                  ...                  ...   
10835                                   Sya9a Maroc - FR               FAMILY   
10836                   Fr. Mike Schmitz Audio Teachings               FAMILY   
10837                             Parkinson Exercices FR              MEDICAL   
10838                      The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10839      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       rating reviews                size     installs  type  price  \
0         4.1     159                 19M      10,000+  Free    0.0   
1         3.9     967                 14M     500,000+  Free    0.0   
2         4.7   87510                8.7M   5,000,000+  Free    0.0   
3         4.5  215644                 25M  50,000,000+  Free    0.0   
4         4.3     967                2.8M     100,000+  Free    0.0   
...       ...     ...                 ...          ...   ...    ...   
10835     4.5      38                 53M       5,000+  Free    0.0   
10836     5.0       4                3.6M         100+  Free    0.0   
10837     NaN       3                9.5M       1,000+  Free    0.0   
10838     4.5     114  Varies with device       1,000+  Free    0.0   
10839     4.5  398307                 19M  10,000,000+  Free    0.0   

      content_rating                     genres      last_updated  \
0           Everyone               Art & Design   January 7, 2018   
1           Everyone  Art & Design;Pretend Play  January 15, 2018   
2           Everyone               Art & Design    August 1, 2018   
3               Teen               Art & Design      June 8, 2018   
4           Everyone    Art & Design;Creativity     June 20, 2018   
...              ...                        ...               ...   
10835       Everyone                  Education     July 25, 2017   
10836       Everyone                  Education      July 6, 2018   
10837       Everyone                    Medical  January 20, 2017   
10838     Mature 17+          Books & Reference  January 19, 2015   
10839       Everyone                  Lifestyle     July 25, 2018   

              current_ver         android_ver  mean_cat_rating day_of_update  
0                   1.0.0        4.0.3 and up         4.358065     miss_date  
1                   2.0.0        4.0.3 and up         4.358065     miss_date  
2                   1.2.4        4.0.3 and up         4.358065     miss_date  
3      Varies with device          4.2 and up         4.358065     miss_date  
4                     1.1          4.4 and up         4.358065     miss_date  
...                   ...                 ...              ...           ...  
10835                1.48          4.1 and up         4.192272     miss_date  
10836                 1.0          4.1 and up         4.192272     miss_date  
10837                 1.0          2.2 and up         4.189143     miss_date  
10838  Varies with device  Varies with device         4.346067     miss_date  
10839  Varies with device  Varies with device         4.094904     miss_date  

[10840 rows x 15 columns]